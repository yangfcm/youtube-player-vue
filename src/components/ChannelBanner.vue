<script setup lang="ts">
const props = defineProps<{
  imageUrl?: string
}>()

</script>

<template>
  <template v-if="imageUrl">
    <div class="app-channel-banner" :style="`background-image: url(${props.imageUrl})`">
    </div>
  </template>
</template>

<style>
  .app-channel-banner {
    /* background-image: "url(v-bind('bannerImageUrl'))"; */
    width: 100%;
    height: 16vw;
    background-size: cover;
    background-repeat: no-repeat;
    background-position: 0 50%;
  }
</style>