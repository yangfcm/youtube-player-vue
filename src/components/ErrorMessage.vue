<script setup lang="ts">
defineProps<{
  message?: string
}>()
</script>

<template>
  <div>
    <v-alert icon="mdi-alert" type="error">{{ message || 'Oops, something went wrong.' }}</v-alert>
  </div>
</template>
