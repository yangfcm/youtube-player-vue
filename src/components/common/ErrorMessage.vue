<template>
  <div>
    <app-blank></app-blank>
    <h3 class="ui red header center aligned">
      <i class="exclamation triangle icon"> </i><slot></slot>
    </h3>
    <app-blank></app-blank>
  </div>
</template>

<script>
import Blank from "./Blank";
export default {
  components: {
    appBlank: Blank,
  },
  props: ["errorMessage"],
};
</script>
<style scoped></style>
