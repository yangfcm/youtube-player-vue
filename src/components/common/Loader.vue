<template>
  <div>
    <app-blank></app-blank>
    <div class="ui active centered inline loader"></div>
    <app-blank></app-blank>
  </div>
</template>

<script>
import Blank from "./Blank";
export default {
  components: {
    appBlank: Blank,
  },
};
</script>

<style scoped></style>
