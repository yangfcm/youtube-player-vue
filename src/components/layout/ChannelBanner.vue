<template>
  <div class="app-banner-container" v-if="channelIntro">
    <div class="app-banner-title-container">
      <div class="app-banner-image">
        <img
          class="ui tiny image circular"
          :src="channelIntro.snippet.thumbnails.medium.url"
          :alt="channelIntro.snippet.title"
        />
      </div>
      <div class="app-banner-title">
        <h2 class="ui header app-banner-title-text">{{ channelIntro.snippet.title }}</h2>
      </div>
    </div>
    <!-- <div>
        <app-subscribe-button-auth
          [channelId]="channelIntro.id"
        ></app-subscribe-button-auth>
    </div>-->
  </div>
</template>

<script>
export default {
  props: ["channelIntro"],
};
</script>

<style scoped>
.app-banner-container {
  margin-top: 1.5rem;
  margin-bottom: 1.5rem;
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.app-banner-title-container {
  display: flex;
}
.app-banner-image {
  margin-right: 1rem;
  min-height: 80px;
  display: flex;
  align-items: center;
}
.app-banner-title {
  display: flex !important;
  align-items: center;
}
.app-banner-title-text {
  display: flex;
  padding-top: 5px;
  padding-bottom: 5px;
}
@media (max-width: 576px) {
  .app-banner-title-text {
    flex-direction: column;
  }
}
@media (max-width: 576px) {
  .app-banner-container {
    flex-direction: column;
  }
}
</style>