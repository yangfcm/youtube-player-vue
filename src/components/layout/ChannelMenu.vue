<template>
  <div class="ui three item menu red stackable">
    <router-link class="item" to="videos" active-class="active">
      <i class="video icon"></i> Videos
    </router-link>
    <router-link class="item" to="playlist" active-class="active">
      <i class="list icon"></i>Playlist
    </router-link>
    <router-link class="item" to="intro" active-class="active">
      <i class="info circle icon"></i>Intro
    </router-link>
  </div>
</template>

<script>
export default {};
</script>

<style scoped>
</style>