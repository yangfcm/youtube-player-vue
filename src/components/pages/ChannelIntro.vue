<template>
  <div>
    <app-loader v-if="!channelIntro && !error"></app-loader>
    <app-error-message v-if="error">{{ error }}</app-error-message>
    <div v-if="channelIntro && !error" class="ui segments">
      <div class="ui segment">
        <p>
          <i class="calendar plus icon"></i>
          Published on {{ channelIntro.snippet.publishedAt | date }}
        </p>
      </div>
      <div class="ui segment">
        <p>
          <i class="file alternate icon"></i>
          Description -
          {{ channelIntro.snippet.description }}
        </p>
      </div>
      <div class="ui segment">
        <p>
          <i class="users icon"></i>
          {{ channelIntro.statistics.subscriberCount | decimal }}
          subscribers
        </p>
      </div>
      <div class="ui segment">
        <p>
          <i class="eye icon"></i>
          {{ channelIntro.statistics.viewCount | decimal }} views
        </p>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState, mapGetters, mapActions } from "vuex";
import ErrorMessage from "../common/ErrorMessage";
import Loader from "../common/Loader";

export default {
  components: {
    appLoader: Loader,
    appErrorMessage: ErrorMessage,
  },
  data: function () {
    return {
      channelIntro: null,
      error: "",
    };
  },
  computed: {
    ...mapState(["channel"]),
    ...mapGetters(["channelErrorMessage"]),
  },
  methods: {
    ...mapActions(["fetchChannelIntro"]),
  },
  async created() {
    const channelId = this.$route.params.id;
    await this.fetchChannelIntro(channelId);
    if (this.channel.channelError) {
      this.error = this.channelErrorMessage;
    } else if (this.channel.channelIntro) {
      if (!this.channel.channelIntro.items) {
        this.$router.push("/not-found");
      }
      this.channelIntro = this.channel.channelIntro.items[0];
      this.error = "";
    }
  },
};
</script>

<style scoped></style>
