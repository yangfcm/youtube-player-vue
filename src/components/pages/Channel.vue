<template>
  <div>
    <app-error-message v-if="error">{{ error }}</app-error-message>
    <div v-else>
      <app-channel-banner :channelIntro="channelIntro"></app-channel-banner>
      <app-channel-menu></app-channel-menu>
      <router-view></router-view>
    </div>
  </div>
</template>

<script>
import { mapState, mapGetters, mapActions } from "vuex";
import ChannelBanner from "../layout/ChannelBanner";
import ChannelMenu from "../layout/ChannelMenu";
import ErrorMessage from "../common/ErrorMessage";
export default {
  components: {
    appChannelBanner: ChannelBanner,
    appChannelMenu: ChannelMenu,
    appErrorMessage: ErrorMessage,
  },
  data: function () {
    return {
      channelIntro: null,
      error: "",
    };
  },
  computed: {
    ...mapState(["channel"]),
    ...mapGetters(["channelErrorMessage"]),
  },
  methods: {
    ...mapActions(["fetchChannelIntro"]),
  },
  async created() {
    const channelId = this.$route.params.id;
    await this.fetchChannelIntro(channelId);
    if (this.channel.channelError) {
      this.error = this.channelErrorMessage;
    } else if (this.channel.channelIntro) {
      if (!this.channel.channelIntro.items) {
        this.$router.push("/not-found");
      }
      this.channelIntro = this.channel.channelIntro.items[0];
      this.error = "";
    }
  },
};
</script>

<style scoped></style>
