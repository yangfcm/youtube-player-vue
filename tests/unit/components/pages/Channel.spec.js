import { mount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import VueRouter from "vue-router";
import { routes } from "@/router";
import Channel from "@/components/pages/Channel";
import ChannelBanner from "@/components/layout/ChannelBanner";
import ChannelMenu from "@/components/layout/ChannelMenu";
import ErrorMessage from "@/components/common/ErrorMessage";
import ChannelVideos from "@/components/pages/ChannelVideos";
import ChannelPlaylist from "@/components/pages/ChannelPlaylist";
import ChannelIntro from "@/components/pages/ChannelIntro";
import Loader from "@/components/common/Loader";
import { channelIntroResponse, channelId } from "../../store/fixtures/channel";

describe("Test Channel page", () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);
  localVue.use(VueRouter);
  const router = new VueRouter({
    routes,
    mode: "history",
  });
  const actions = {
    fetchChannelIntro: jest.fn(),
    fetchChannelSubscription: jest.fn(),
    fetchVideos: jest.fn(),
    searchVideos: jest.fn(),
  };
  const store = new Vuex.Store({
    state: {
      channel: {
        channelIntro: channelIntroResponse,
        channelError: null,
      },
      video: {
        videoError: null,
      },
      search: {
        searchError: null,
      },
    },
    getters: {
      channelErrorMessage: jest.fn(() => ""),
      signedIn: jest.fn(() => false),
    },
    actions,
    data() {
      return {
        channelIntro: null,
        error: "",
      };
    },
  });
  let wrapper;
  beforeEach(() => {
    wrapper = mount(Channel, {
      store,
      localVue,
      router,
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should render ErrorMessage if error occurs", async () => {
    const errorMessage = "mock error message";
    wrapper.setData({
      error: errorMessage,
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.findComponent(ErrorMessage).exists()).toBe(true);
  });

  it("should render ChannelBanner and pass prop props to it", () => {
    expect(wrapper.findComponent(ChannelBanner).exists()).toBe(true);
    expect(wrapper.findComponent(ChannelBanner).props("channelIntro")).toEqual(
      channelIntroResponse.items[0]
    );
  });

  it("should render ChannelMenu component", () => {
    expect(wrapper.findComponent(ChannelMenu).exists()).toBe(true);
  });

  it("should render ChannelVideos component for route /channel/channelId/videos", async () => {
    router.push(`/channel/${channelId}/videos`);
    await wrapper.vm.$nextTick();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.findComponent(ChannelVideos).exists()).toBe(true);
  });
});
