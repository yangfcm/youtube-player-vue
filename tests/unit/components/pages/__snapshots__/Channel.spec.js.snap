// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Channel page should render ChannelVideos component for route /channel/channelId/videos 1`] = `
<div>
  <div>
    <div class="app-banner-container">
      <div class="app-banner-title-container">
        <div class="app-banner-image"><img src="https://yt3.ggpht.com/a/AATXAJzt3HlB60RSLIwrWgXUfKLZeo0M_rC34JMcM0BPZw=s240-c-k-c0xffffffff-no-rj-mo" alt="BBC Learning English" class="ui tiny image circular"></div>
        <div class="app-banner-title">
          <h2 class="ui header app-banner-title-text">BBC Learning English</h2>
        </div>
      </div>
      <div>
        <div style="padding-top: 8px;">
          <!---->
        </div>
      </div>
    </div>
    <div class="ui three item menu red stackable"><a href="/videos" class="item"><i class="video icon"></i> Videos
      </a> <a href="/playlist" class="item"><i class="list icon"></i>Playlist
      </a> <a href="/intro" class="item"><i class="info circle icon"></i>Intro
      </a></div>
    <div>
      <div class="app-banner-container">
        <div class="app-banner-image"><i class="user secret icon huge"></i>
          <!---->
        </div>
        <div class="app-banner-title">
          <h2 class="ui header app-banner-title-text"><span style="margin-right: 3px;">Welcome
        
      </span></h2>
        </div>
      </div>
      <div class="ui three item menu red stackable"><a href="/" aria-current="page" class="item router-link-exact-active active"><i class="thumbs up outline icon"></i> Recommend
        </a> <a href="/subscription" class="item"><i class="user plus icon"></i> Subscription
        </a> <a href="/playlist" class="item"><i class="list icon"></i> Playlist</a></div>
      <div>
        <div class="ui hidden divider"></div>
        <div class="ui active centered inline loader"></div>
        <div class="ui hidden divider"></div>
      </div>
      <!---->
      <!---->
    </div>
  </div>
</div>
`;
